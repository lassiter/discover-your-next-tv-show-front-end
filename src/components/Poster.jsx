import React, { Component } from 'react'
import styled from 'styled-components'
import { Link } from 'react-router-dom'

const InternalPoster = styled(Link)`
  display: block;
  position: relative;
  min-width: 300px;
  min-height: 450px;
  width: min-content;
  margin: auto;
`

const PosterImage = styled.img`
  max-width: 300px;
  min-width: 150px;
  min-height: 450px;
`

const PosterHeader = styled.h3`
  display: block;
  font-size: 2em;
  color: white;

  margin-block-start: 0em;
  margin-block-end: 0em;
  margin-inline-start: 0px;
  margin-inline-end: 0px;
  font-weight: bold;
`
const PosterOverview = styled.p`
  display: block;
  font-size: 1em;
  color: white;
  margin-block-start: 0em;
  margin-block-end: 0em;
  margin-inline-start: 0px;
  margin-inline-end: 0px;
`

const HoverContainer = styled.div`
  &.null-poster {
    background: linear-gradient(180deg, rgba(0,0,0,1), rgba(0,0,0,0) 100%);
    width: 300px;
    height: 450px;
    position: absolute;
    top: 0;
  }
  &.hidden {
    display: none;
  }
  &:not(.null-poster) {
    background: linear-gradient(180deg, rgba(0,0,0,1), rgba(0,0,0,0) 100%);
    width: 100%;
    height: 100%;
    position: absolute;
    top: 0;
  }
`

export default class Poster extends Component {
  constructor(props){
    super(props)
    this.hoverContainerRef = React.createRef();
    this.state = {
      className: "hidden",
      hoverIsActive: false
    }
    this.handleHover = () => {
      if (!this.state.hoverIsActive) {
        this.setState({
          hoverIsActive: true,
          className: ""
        })
      } else {
        this.setState({
          hoverIsActive: null,
          className: "hidden"
        })
      }
    }
  }
  componentDidCatch(error, info) {
    // You can also log the error to an error reporting service
    console.log(error, info);
  }
  render() {
    if (this.props.showData.overview === "" && this.props.showData.poster_path === null) {
      return null
    }
    if (this.props.showData.poster_path === null) {
      return (
        <InternalPoster 
          alt={this.props.showData.name}
          onMouseEnter={this.handleHover}
          onMouseLeave={this.handleHover}
          to={{
            pathname: `/${this.props.showData.name.split(" ").join('_')}`,
            state: {
              show: this.props.showData
            }
          }}
        >
          <HoverContainer className={"null-poster"} ref={this.hoverContainerRef}>
            <PosterHeader>{this.props.showData.name}</PosterHeader>
            <PosterOverview>{this.props.showData.overview}</PosterOverview>
          </HoverContainer>
        </InternalPoster>
      )
    } else {
      return (
        <InternalPoster 
          alt={this.props.showData.name}
          onMouseEnter={this.handleHover}
          onMouseLeave={this.handleHover}
          to={{
            pathname: `/${this.props.showData.name.split(" ").join('_')}`,
            state: {
              show: this.props.showData
            }
          }}
        >
          <PosterImage src={`https://image.tmdb.org/t/p/w500${this.props.showData.poster_path}`}/>
          <HoverContainer className={this.state.className} ref={this.hoverContainerRef}>
            <PosterHeader>{this.props.showData.name}</PosterHeader>
            <PosterOverview>{this.props.showData.overview}</PosterOverview>
          </HoverContainer>
        </InternalPoster>
      )
    }
  }
}
