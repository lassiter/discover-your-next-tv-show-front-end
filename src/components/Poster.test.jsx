import React from 'react';
import Poster from './Poster';
import renderer from 'react-test-renderer';
import jsonTestData from '../testJSON/got_sample_data.json'
import { StaticRouter } from 'react-router'


it('renders correctly', () => {
  const voteAverage = 8.2
  const voteCount = 100
  const tree = renderer
    .create(
    <StaticRouter location="/">
      <Poster showData={JSON.parse(jsonTestData)}/>
    </StaticRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});