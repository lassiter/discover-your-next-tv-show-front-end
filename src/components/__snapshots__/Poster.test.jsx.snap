// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<a
  alt="Game of Thrones"
  className="sc-bdVaJa kZDaAF"
  href="/Game-of-Thrones"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <img
    className="sc-bwzfXH keFRXT"
    src="https://image.tmdb.org/t/p/w500/u3bZgnGQ9T01sWNhyveQz0wH0Hl.jpg"
  />
  <div
    className="hidden sc-ifAKCX yjWSy"
  >
    <h3
      className="sc-htpNat lctXZJ"
    >
      Game of Thrones
    </h3>
    <p
      className="sc-bxivhb fJKucG"
    >
      Seven noble families fight for control of the mythical land of Westeros. Friction between the houses leads to full-scale war. All while a very ancient evil awakens in the farthest north. Amidst the war, a neglected military order of misfits, the Night's Watch, is all that stands between the realms of men and icy horrors beyond.
    </p>
  </div>
</a>
`;
